src/python/pants/util/objects.py

```diff
@@ -132,13 +132,16 @@ def __eq__(self, other):
     def __ne__(self, other):
       return not (self == other)
 
+    # NB: it appears that in Python 3, whenever __eq__ is overridden, __hash__() must also be
+    # explicitly implemented, otherwise Python will raise "unhashable type". It's not clear whether
+    # or where this behavior is documented.
```

Comment:
This is correct. I ran into this issue very early on in the project IIRC :/ 

Note that this behavior is documented, however. See https://docs.python.org/3/reference/datamodel.html#object.__hash__. Would be good to update comment to include this link.